engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java

```diff
@@ -191,6 +192,13 @@ public void createInstance(final ElementInstance instance) {
     elementInstanceColumnFamily.insert(elementInstanceKey, instance);
     parentChildColumnFamily.insert(parentChildKey, DbNil.INSTANCE);
     variableState.createScope(elementInstanceKey.getValue(), parentKey.inner().getValue());
+
+    final var recordValue = instance.getValue();
+    if (recordValue.getBpmnElementType() == BpmnElementType.PROCESS) {
+      processDefinitionKey.wrapLong(recordValue.getProcessDefinitionKey());
+      processInstanceKeyByProcessDefinitionKeyColumnFamily.insert(
+          processInstanceKeyByProcessDefinitionKey, DbNil.INSTANCE);
```

Comment:
💭 If we make `processDefinitionKey` a foreign key to the process definition, we'd be able to completely avoid the situation that we saw in:
- https://github.com/camunda/zeebe/issues/13254

I previously [wrote about that](https://github.com/camunda/zeebe/issues/13254#issuecomment-1617550522):
>If we had such a check, the corrupted state would've been avoided. The affected partition would've been DEAD but recoverable after a patch. Such a check would avoid all errors (including the unknown ones) related to process instance creation where the process definition is actually not stored in state.
>
>Keeping track of this foreign key relation would come at the cost of performance (two additional IO operations on process instance creation: one to store the additional entry in a new column family, and one to check the foreign key relation when the instance is being created in the state).

But, if we have this in place anyways to support #9576, then that cost is significantly lower. Only the foreign key assertion (lookup) on insert would be needed. Let's consider this.

@remcowesterhoud If you agree, we could move this to a separate issue. That allows us to benchmark the difference to this change.

🔧 We should benchmark the current change anyways as well.

Reply:
> If we make processDefinitionKey a foreign key to the process definition, we'd be able to completely avoid the situation that we saw in:

I get what you're saying, but this also means our `ProcessState` and `ElementInstanceState` would be aware of each others existence. I'm not quite sure how I feel about that.

Reply:
That's a fair point, but the [timers have a foreign key to element instances](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbTimerInstanceState.java#L45-L49), and incidents have a foreign key [to element instances](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbIncidentState.java#L57) and [to jobs](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbIncidentState.java#L65). And similarly, we can add a relation from process instance (element instance of type process) to process definition. I think this is reasonable for relational data constraints.

Reply:
I wasn't aware we already had such dependencies. In that case I don't see why not 🤷 Would you mind creating the issue? I think you have more context on the original issue.